File: 2019/13/solution.ts
import * as fs from "fs";
import * as path from "path";
import intcodeGenerator from "../intcode/intcode";

const puzzleInput = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split(",")
  .map(Number);

const arcadeCabinet = intcodeGenerator(puzzleInput);

let part1 = 0;

while (true) {
  if (arcadeCabinet.next().done) break;
  arcadeCabinet.next();
  const tileRet = arcadeCabinet.next();
  if (tileRet.value === 2) part1++;
}

console.log(part1);

const freePlayInput = puzzleInput.slice();
freePlayInput[0] = 2;

const freePlayArcade = intcodeGenerator(freePlayInput);
let move = 0;
let xBall = 0;
let xPaddle = 0;
let part2 = 0;
while (true) {
  let xRet: IteratorResult<number, number>;
  while (true) {
    xRet = freePlayArcade.next(move);
    if (xRet.done || xRet.value === -9999) break;
    const yRet = freePlayArcade.next();
    const tileRet = freePlayArcade.next();
    if (xRet.value === -1 && yRet.value === 0) {
      part2 = tileRet.value;
    }
    if (tileRet.value === 4) {
      xBall = xRet.value;
    } else if (tileRet.value === 3) {
      xPaddle = xRet.value;
    }
  }
  if (xRet.done) break;
  if (xBall < xPaddle) move = -1;
  else if (xBall > xPaddle) move = 1;
  else move = 0;
}
console.log(part2);

File: 2019/13/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"sync"

	"github.com/weitecklee/adventofcode/2019/intcode"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), ","))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func parseInput(data []string) []int {
	numbers := make([]int, 0, len(data))
	for _, s := range data {
		if n, err := strconv.Atoi(s); err != nil {
			panic(err)
		} else {
			numbers = append(numbers, n)
		}
	}
	return numbers
}

func part1(puzzleInput []int) int {
	var wg sync.WaitGroup
	inChan := make(chan int)
	defer close(inChan)
	outChan := make(chan int)
	res := 0
	arcade := intcode.NewIntcodeProgram(puzzleInput, inChan, outChan, &wg)
	var tileId int
	wg.Add(1)
	go arcade.Run()
	for range outChan {
		<-outChan
		tileId = <-outChan
		if tileId == 2 {
			res++
		}
	}
	return res
}

func part2(puzzleInput []int) int {
	var wg sync.WaitGroup
	inChan := make(chan int)
	defer close(inChan)
	outChan := make(chan int)
	puzzleInput[0] = 2
	arcade := intcode.NewIntcodeProgram(puzzleInput, inChan, outChan, &wg)
	var x, tileId, score, ballPos, paddlePos int
	wg.Add(1)
	go arcade.Run()
	for {
		for x = range outChan {
			if x == intcode.REQUESTSIGNAL || x == intcode.ENDSIGNAL {
				break
			}
			<-outChan
			tileId = <-outChan
			if tileId == 3 {
				paddlePos = x
			} else if tileId == 4 {
				ballPos = x
			} else if x == -1 {
				score = tileId
			}
		}
		if x == intcode.ENDSIGNAL {
			break
		}
		if ballPos > paddlePos {
			inChan <- 1
		} else if ballPos < paddlePos {
			inChan <- -1
		} else {
			inChan <- 0
		}
	}
	return score
}
